internal/nri/container.go

```diff
@@ -0,0 +1,58 @@
+package nri
+
+import (
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+
+	nri "github.com/containerd/nri/pkg/adaptation"
+)
+
+// Container interface for interacting with NRI.
+type Container interface {
```

Comment:
I guess what I don't understand is why these interfaces are introduced here. I get it they are internal to cri-o.

Is there going to be more than 1 implementation? Or, in general, what is the reason to have these interfaces?

Reply:
I tried to keep the containerd NRI integration and the crio one structurally as close to each other as possible. One reason was to lower the cost of context-switching between them and ease maintenance. But another reason was that I still hope that once the dust settles, I'd be able to eventually further generalize these bits enough to move them over to the NRI repo as a package, and then consume it from there in both runtimes. I just couldn't give that a serious try at this phase.

The bits here are mostly agnostic to the internal runtime-specific implementation/representation of pods and containers, but need to be able to introspect and manipulate these, to interact with and interface NRI plugins. The interfaces here are for that. The rest of the patch set uses this internal package to hook NRI into CRI request processing.  

On the containerd side it is even possible to have multiple sandbox and container implementations but support for that is exactly what I gutted out from the code here in cri-o.